flink-core/src/test/java/org/apache/flink/api/common/operators/util/RandomSamplerTest.java

```diff
@@ -0,0 +1,425 @@
+	}
+	
+	@Test
+	public void testBernoulliSamplerDistribution() {
+		verifyBernoulliSampler(0.01d);
+		verifyBernoulliSampler(0.05d);
+		verifyBernoulliSampler(0.1d);
+		verifyBernoulliSampler(0.5d);
+	}
+	
+	@Test
+	public void testPoissonSamplerDistribution() {
+		verifyPoissonSampler(0.01d);
+		verifyPoissonSampler(0.05d);
+		verifyPoissonSampler(0.1d);
+		verifyPoissonSampler(0.5d);
+	}
+	
+	@Test
+	public void testReservoirSamplerSampledSize() {
+		verifySamplerFixedSampleSize(1, true);
+		verifySamplerFixedSampleSize(10, true);
+		verifySamplerFixedSampleSize(100, true);
+		verifySamplerFixedSampleSize(1234, true);
+		verifySamplerFixedSampleSize(9999, true);
+		verifySamplerFixedSampleSize(20000, true);
+		
+		verifySamplerFixedSampleSize(1, false);
+		verifySamplerFixedSampleSize(10, false);
+		verifySamplerFixedSampleSize(100, false);
+		verifySamplerFixedSampleSize(1234, false);
+		verifySamplerFixedSampleSize(9999, false);
+	}
+	
+	@Test
+	public void testReservoirSamplerSampledSize2() {
+		RandomSampler<Double> sampler = new ReservoirSamplerWithoutReplacement<Double>(20000);
+		Iterator<Double> sampled = sampler.sample(source.iterator());
+		assertTrue("ReservoirSamplerWithoutReplacement sampled output size should not beyond the source size.", getSize(sampled) == SOURCE_SIZE);
+	}
+	
+	@Test
+	public void testReservoirSamplerDuplicateElements() {
+		verifyRandomSamplerDuplicateElements(new ReservoirSamplerWithoutReplacement<Double>(100));
+		verifyRandomSamplerDuplicateElements(new ReservoirSamplerWithoutReplacement<Double>(1000));
+		verifyRandomSamplerDuplicateElements(new ReservoirSamplerWithoutReplacement<Double>(5000));
+	}
+	
+	@Test
+	public void testReservoirSamplerWithoutReplacement() {
+		verifyReservoirSamplerWithoutReplacement(100, false);
+		verifyReservoirSamplerWithoutReplacement(500, false);
+		verifyReservoirSamplerWithoutReplacement(1000, false);
+		verifyReservoirSamplerWithoutReplacement(5000, false);
+	}
+	
+	@Test
+	public void testReservoirSamplerWithReplacement() {
+		verifyReservoirSamplerWithReplacement(100, false);
+		verifyReservoirSamplerWithReplacement(500, false);
+		verifyReservoirSamplerWithReplacement(1000, false);
+		verifyReservoirSamplerWithReplacement(5000, false);
+	}
+
+	@Test
+	public void testReservoirSamplerWithMultiSourcePartitions1() {
+		initSourcePartition();
+
+		verifyReservoirSamplerWithoutReplacement(100, true);
+		verifyReservoirSamplerWithoutReplacement(500, true);
+		verifyReservoirSamplerWithoutReplacement(1000, true);
+		verifyReservoirSamplerWithoutReplacement(5000, true);
+	}
+
+	@Test
+	public void testReservoirSamplerWithMultiSourcePartitions2() {
+		initSourcePartition();
+
+		verifyReservoirSamplerWithReplacement(100, true);
+		verifyReservoirSamplerWithReplacement(500, true);
+		verifyReservoirSamplerWithReplacement(1000, true);
+		verifyReservoirSamplerWithReplacement(5000, true);
+	}
+	
+	private void verifySamplerFixedSampleSize(int numSample, boolean withReplacement) {
```

Comment:
Having some comments for some of the ``verify[..]`` functions explaining what we verify at each would help with code understanding.